src/test/rpc/ValidatorRPC_test.cpp

```diff
@@ -0,0 +1,332 @@
+                }
+            }
+        }
+    }
+
+    void
+    testStaticUNL()
+    {
+        using namespace test::jtx;
+
+        std::set<std::string> const keys = {
+            "n949f75evCHwgyP4fPVgaHqNHxUVN15PsJEZ3B3HnXPcPjcZAoy7",
+            "n9MD5h24qrQqiyBC8aeqqCWvpiBiYQ3jxSr91uiDvmrkyHRdYLUj"};
+        Env env{
+            *this,
+            envconfig([&keys](std::unique_ptr<Config> cfg) {
+                for (auto const& key : keys)
+                    cfg->section(SECTION_VALIDATORS).append(key);
+                return cfg;
+            }),
+        };
+
+        // Server info reports maximum expiration since not dynamic
+        {
+            auto const jrr = env.rpc("server_info")[jss::result];
+            BEAST_EXPECT(
+                jrr[jss::info][jss::validator_list_expires] ==
+                to_string(NetClock::time_point::max()));
+        }
+        // All our keys are in the response
+        {
+            auto const jrr = env.rpc("validator_lists")[jss::result];
+            BEAST_EXPECT(
+                jrr[jss::validator_list_expires] ==
+                to_string(NetClock::time_point::max()));
+            BEAST_EXPECT(jrr[jss::validation_quorum].asUInt() == keys.size());
+            BEAST_EXPECT(jrr[jss::validator_keys].size() == keys.size());
+            std::set<std::string> foundKeys;
+            for (auto const& jKeys : jrr[jss::validator_keys])
+            {
+                foundKeys.insert(jKeys[jss::pubkey_validator].asString());
+            }
+            BEAST_EXPECT(foundKeys == keys);
+        }
+        // No validator sites configured
+        {
+            auto const jrr = env.rpc("validator_sites")[jss::result];
+            BEAST_EXPECT(jrr[jss::validator_sites].size() == 0);
+        }
+    }
+
+    void
+    testDynamicUNL()
+    {
+        using namespace test::jtx;
+        using endpoint_type = boost::asio::ip::tcp::endpoint;
+        using address_type = boost::asio::ip::address;
+
+        auto toStr = [](PublicKey const& publicKey) {
+            return toBase58(TokenType::TOKEN_NODE_PUBLIC, publicKey);
+        };
+
+        // Publisher manifest/signing keys
+        auto const publisherSecret = randomSecretKey();
+        auto const publisherPublic =
+            derivePublicKey(KeyType::ed25519, publisherSecret);
+        auto const publisherSigningKeys = randomKeyPair(KeyType::secp256k1);
+        auto const manifest = makeManifestString(
+            publisherPublic,
+            publisherSecret,
+            publisherSigningKeys.first,
+            publisherSigningKeys.second,
+            1);
+
+        // Validator keys that will be in the published list
+        std::vector<PublicKey> keys = {randomNode(), randomNode()};
+        std::set<std::string> expectedKeys;
+        for (auto const& key : keys)
+            expectedKeys.insert(toStr(key));
+
+        // Publisher site information
+        std::uint16_t constexpr port = 7475;
```

Comment:
for tests particularly, I would opt for OS port selection - it helps avoids spurious failures related to ports in use. You can use port `0` in for the bind endpoint and then provide a way on `TrustedPublisherServer` to get `acceptor_.local_endpoint()` after `listen` has been called, which you then pass to clients (probably the simplest approach is just make `acceptor_` public).